### PR TITLE
Docs. Closes #4

### LINKED ISSUE #4: Add proper godoc
Once #1 is solved, and it is validated that no further API changes are required

### PR DESCRIPTION
**Pull request by ribasushi**
This is a lot of text (and code). This is essentially my first-ever standalone golang-based contribution.

Roast me! Anything goes: documentation style nits, naming nits, code nits. The library does seem to work flawlessly ( 48h very high cncurrency spin so far with no deadlocks/overwrites ).

/cc @warpfork as well ( can't add you as reviewer for some reason )

### CODE REVIEW

File: qringbuf.go [hunk 1]
```diff
@@ -1,3 +1,166 @@
...
+// io.ReadAtLeast() over a pre-allocated ring-buffer, populated asynchronously
+// by a standalone goroutine. It is primarily designed for processing a series
+// of arbitrary streams each comprised of variable-length records.
+//
+// Specifically an object of this package makes the following guarantees:
+//  • Memory is allocated only at construction time, never during streaming
+//  • StartFill() spawns off a single goroutine (collector) which terminates when:
+//    ◦ It reaches readLimit or receives an expected io.EOF from the wrapped Reader
+//    ◦ Immediately at any other error
+//    ◦ Af the next chance to evaluate a signal received from StopFill()
```
Comment (warpfork) [inline qringbuf.go:11]: s/Af/At/

File: qringbuf.go [hunk 2]
```diff
@@ -1,3 +1,168 @@
...
+// of arbitrary streams each comprised of variable-length records.
+//
+// Specifically an object of this package makes the following guarantees:
+//  • Memory is allocated only at construction time, never during streaming
+//  • StartFill() spawns off a single goroutine (collector) which terminates when:
+//    ◦ It reaches readLimit, if one was supplied
+//    ◦ It receives any error from the wrapped reader (including io.EOF)
+//    ◦ It had a chance to evaluate a signal received from StopFill()
+//      (most io.Read() operations are not cancelable, and may block forever)
+//  • Every call to NextRegion() blocks until it can return:
```
Comment (dirkmc) [inline qringbuf.go:13]: Nit: It wasn't clear to me on the first reading that NextRegion() takes a parameter

File: qringbuf.go [hunk 3]
```diff
@@ -1,3 +1,168 @@
...
+//    ◦ A *Region object representing a contiguous slice of at least MinRegion bytes
+//    ◦ A smaller "trailing" *Region and an io.EOF or ErrCollectorStopped
```
Comment (dirkmc) [inline qringbuf.go:15]: I'm not sure what a "trailing" region means here. It might make sense to give a small example of the methods on the API and how they're typically used before the guarantees section, to give the reader context

File: qringbuf.go [hunk 4]
```diff
@@ -1,3 +1,168 @@
...
+//    ◦ A nil-Region and any other error
+//  • *Region.Bytes() is always a slice of the underlying buffer, no data copying takes place
+//  • Data backing a *Region is guaranteed to remain intact provided:
```
Comment (dirkmc) [inline qringbuf.go:18]: nit: does intact mean "not get overwritten"?

File: qringbuf.go [hunk 5]
```diff
@@ -1,3 +1,168 @@
...
+//    ◦ NextRegion() has not been called again allowing writes into the buffer
+//    ◦ *Region.Reserve() was invoked, which blocks writes until a *Region.Release()
+//
+// Examples
+//
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
+//  …
+//  readLimit = int64( stopReadingAfterThatManyBytes )
```
Comment (dirkmc) [inline qringbuf.go:28]: Any reason to use int64 here instead of uint64?

File: qringbuf.go [hunk 6]
```diff
@@ -1,3 +1,168 @@
...
+//  startErr := qrb.StartFill( readLimit )
+//  …
+//  var available, processed int
+//  for {
+//      reg, streamErr := qrb.NextRegion( available - processed )
+//      if reg == nil {
```
Comment (dirkmc) [inline qringbuf.go:34]: Usually in go I'd expect to check `if streamErr != nil`, is there a reason to instead check `if reg == nil`?

File: qringbuf.go [hunk 7]
```diff
@@ -1,3 +1,168 @@
...
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
+//  …
+//  readLimit = int64( stopReadingAfterThatManyBytes )
+//  startErr := qrb.StartFill( readLimit )
+//  …
+//  var available, processed int
+//  for {
+//      reg, streamErr := qrb.NextRegion( available - processed )
```
Comment (dirkmc) [inline qringbuf.go:33]: I'd have to think a bit about why the parameter to NextRegion is `available - processed`. Could you add a comment explaining this?

File: qringbuf.go [hunk 8]
```diff
@@ -1,3 +1,168 @@
...
+// Here is an illustration of a qringbuf object lifecycle initialized with
+//  {
+//    BufferSize: 64,
+//    MinRegion:  16,
+//    MinRead:    8,
+//    MaxCopy     24,
+//  }
+// ·
+//
+//  † C is the collector position: the *end* of the most recent read from the underlying io.Reader
```
Comment (dirkmc) [inline qringbuf.go:102]: Cool diagrams 👍 

I think this section may be a little clearer if it had an example showing the various function calls, and then the diagrams refer to those line numbers.

**Review by dirkmc — COMMENTED** [review]
Nice diagrams :)

I just had a look at the documentation sections to begin with, mostly nits there, I'll take a look at the code sections in a little bit

File: qringbuf.go (hunk 2 above)
Comment (ribasushi) [reply qringbuf.go]: Addressed as https://github.com/ipfs/go-qringbuf/pull/6/commits/297eacd4921ad6
From commit:
> Did not use actual variables - when rendered I find it too distracting
> Happy to redo if a better suggestion comes along

File: qringbuf.go (hunk 5 above)
Comment (ribasushi) [reply qringbuf.go:28]: I modeled this on the entirety of io.*, which never uses uint ( actually stdlib seems to avoid stdlib as a whole): https://golang.org/pkg/io
I assume this is to make it easier to detect underflows - just a bit too much subtracting and you are into "impossible negative number" territory, as opposed to 2^64-1

File: qringbuf.go (hunk 8 above)
Comment (ribasushi) [reply qringbuf.go:102]: Thanks! Took way longer than I initially thought it will, but actually uncovered an inconsistency while I was drawing them, so all is well.

I am not entirely sure what you are suggesting. Perhaps discussing this voice some time today would help?

File: qringbuf.go (hunk 7 above)
Comment (ribasushi) [reply qringbuf.go:36]: Attempted to address in https://github.com/ipfs/go-qringbuf/pull/6/commits/fbfe0db2fda82 Let me know if not sufficiently clear...

File: README.md [hunk 9]
```diff
@@ -4,9 +4,14 @@ q(uantized)ringbuf(fer)
...
-clients require protection from under-reads in addition to all reads being
-contiguous byte ranges.
+[Bip Buffer][1] and [spsc-bip-buffer/bbqueue][2]. It provides a
+concurrency-friendly, zero-copy abstraction of [io.ReadAtLeast()][3] over a
+pre-allocated ring-buffer, populated asynchronously by a standalone goroutine.
+Refer to the [implementation-notes diagrams][4] to get a quick overview of
+how this works in practice.
+
+This library is primarily designed for processing a series of arbitrary
+streams, each comprised of variable-length records.
```
[the remark this answers is not in the document]
Comment (ribasushi) [reply README.md:14]: I was trying to put in one sentence the entire description of StartFill. Here is a punt of "go read there": https://github.com/ipfs/go-qringbuf/pull/6/commits/67a3224d1a78c
I can't think of a good way to distill this better. Suggestions welcome!

File: qringbuf.go (hunk 3 above)
Comment (ribasushi) [reply qringbuf.go]: See https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

File: qringbuf.go (hunk 4 above)
Comment (ribasushi) [reply qringbuf.go]: See https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

File: qringbuf.go (hunk 6 above)
Comment (ribasushi) [reply qringbuf.go]: TLDR: this did end up making me change code, but the can of worms is strong with that one: https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

Now the long version: This is how I wrote go at first too, and then got legit burned in some unrelated code, realizing **checking for error first/only is incorrect**. From https://golang.org/pkg/io/#Reader :

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. ... Callers **should always process the n > 0 bytes returned before considering the error err**. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors. 

In the example you are asking about, consider that `io.EOF` will start showing up the moment the collector reaches it ( and shuts down ). One will still cycle over the remaining, probably very large region, potentially multiple times requesting less and less of the remaining data: each of these calls will indicate `by the way, nothing extra is coming anymore`, but you are not in an actual "stop work" error condition.

On the other hand I did it wrong anyway: as it was, if the collector "eats" some data, and then the underlying stream errors: the data is now locked in the buffer without a way to get it out. So reworked as https://github.com/ipfs/go-qringbuf/pull/6/commits/4d65af211de

Comment (dirkmc) [reply qringbuf.go]: Ok makes sense - I think the important thing is to make clear any deviation from idiomatic patterns and the example comments & explanation you added address that 👍

File: qringbuf.go (hunk 8 above)
Comment (dirkmc) [reply qringbuf.go:115]: What I mean is that the numbered items seem to be describing a piece of code:
⓪ Buffer initialized, StartFill(…) is called ...
① NextRegion(0) is blocked until ...

So it might be clearer if there's a concrete example of that code to compare with:
Line 3: Buffer initialized, StartFill(…) is called ...
Line 5: NextRegion(0) is blocked until ...

Happy to talk on the phone if that's not clear :)

Comment (ribasushi) [reply qringbuf.go:125]: Ohhhhh I see. This... wouldn't be... something that can be shown in docs, as the diagrams show the "tricky bits" which generally arise quite seldomly if ever. In fact the only way I managed to test all this is by essentially fuzzing the library by spinning it on an overloaded machine starved of CPU time.

I added a bit more prose in https://github.com/ipfs/go-qringbuf/commit/cb4888ddf8db, below is what it looks like visually. If I misunderstood what you are suggesting: let's chat after the launch meet
image

Comment (dirkmc) [reply qringbuf.go:125]: Ok no worries, I don't think you need the disclaimer - it was just a suggestion to add some context to the diagrams, but I think it's fine as is 👍

**Review by dirkmc — APPROVED** [review]

File: qringbuf.go (hunk 8 above)
Comment (ribasushi) [reply qringbuf.go:125]: No, the disclaimer is good to have - if anything it puts things in perspective. So I'll def keep it
Thanks for all your suggestions, things are *WAY* better now ;)

**Comment by ribasushi** [comment]
I will keep this open for a bit longer if anyone else wants to chime in. Will merge Wed the 6th if I don't hear any other API suggestions.

**Comment by dirkmc** [comment]
My review was just for this PR - but it occurs to me now that you may have also wanted a review of all the code, is that right?

**Comment by ribasushi** [comment]
@dirkmc it certainly won't hurt, yes ;) But also if time is short - don't sweat it!

**Comment by ribasushi** [comment]
@dirkmc mainly I am interested in an in-depth "surface API" review. That is: when the code does something dumb, we can fix it. API - more difficult.

File: qringbuf.go [hunk 10]
```diff
@@ -1,3 +1,191 @@
...
+//
+// Unlike io.ReadAtLeast, errors from the underlying reader are always made
+// available on NextRegion(…). As with the standard io.* semantics an error can
+// be returned together with a result. One should always check whether the
+// *Region return value is nil first, before processing the error,
+//
+// Changes of the emitter and collector positions are protected by a mutex on
+// the qringbuf object. Any call accessing the buffer state will block until
+// this lock can be obtained. One can also lock the mutex if the collector
+// goroutine needs to be paused for whatever reason.
```
Comment (Stebalien) [inline qringbuf.go:94]: Avoid exposing mutexes like this. It makes internal locking patterns a part of the API.

File: qringbuf.go [hunk 11]
```diff
@@ -255,16 +483,59 @@ func (qrb *QuantizedRingBuffer) signalCond(c chan<- struct{}) {
...
+//  }
+func (qrb *QuantizedRingBuffer) Buffered() int {
+	defer qrb.Unlock()
+	qrb.Lock()
+	return qrb.cPos - (qrb.ePos + qrb.curRegionSize)
+}
+
+// NextRegion returns a *Region object representing a portion of the underlying
+// stream. One can explicitly request overlapping *Region's by supplying the
+// number of bytes to "step back" (use 0 for "just give me what's next"). This
```
Comment (Stebalien) [inline qringbuf.go:505]: I expect seeking backwards will be common, right? In that case, how about re-defining `NextRegion` as `NextRegion(offset)` where `offset` seeks _forward_ from the start of the previous region. Calling this with `0` would just return the previously read region. That should lead to fewer arithmetic errors.

File: qringbuf.go [hunk 12]
```diff
@@ -25,8 +221,21 @@ type Region struct {
...
+func (r *Region) Size() int { return r.size }
+
+// Bytes returns a slice of the underlying ring buffer. One should take care
+// to copy or finish using the returned slice before making another call to
+// NextRegion(…). If Reserve() has been invoked, then the slice is guaranteed
+// to remain intact before the corresponding Release() call.
 func (r *Region) Bytes() []byte { return r.qrb.buf[r.offset : r.offset+r.size] }
+
+// SubRegion is analogous to re-slicing. Supplying offset/length values causing
+// an out of bounds re-slice results in log.Panic(). As a special case one can
```
Comment (Stebalien) [inline qringbuf.go:236]: If you want to be _really_ fast and sneaky, you could say that this is undefined behavior. If you did that, you'd be able to stick used regions into a `sync.Pool`.

**Review by Stebalien — COMMENTED** [review]
Nice! I actually like this API overall.

File: qringbuf.go [hunk 13]
```diff
@@ -545,8 +824,31 @@ func (qrb *QuantizedRingBuffer) Stop() (didResultInStop bool) {
 	return
 }
 
-func (qrb *QuantizedRingBuffer) Restart(readLimit int64) error {
-
+// StartFill is called at the start of every stream-cycle. If readLimit is 0
```
[the remark this answers is not in the document]
Comment (ribasushi) [reply qringbuf.go:827]: See https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L833-L850
Essentially qringbuf does not care what your offset is within the reader. It just... goes forward.

This is also why I explicitly expose the mutex as stable part of the API - it allows me to "suspend everything" and jump way forward if I need to... though it now occurs to me that I need to remove the locking from Buffered() in order to make this workable.

Back to your questions above:
- Yes, otherwise NextRegion will say "dunno what to give you: a collector is not running"
- If you say 0, the only thing that changes is when the collector exits ( io.EOF ). Notably:
https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L890-L894 in combination with https://github.com/ipfs/go-qringbuf/blob/e7dae312664e482074561c7c913605c6cdc58230/qringbuf.go#L769-L779

Suggestion how to make this... clearer?

[the remark this answers is not in the document]
Comment (Stebalien) [reply qringbuf.go:827]: > This is also why I explicitly expose the mutex as stable part of the API - it allows me to "suspend everything" and jump way forward if I need to... though it now occurs to me that I need to remove the locking from Buffered() in order to make this workable.

Why do you need to _suspend_ it? That seems very strange. If you need a way to "pause", just expose a separate method and automatically resume when the user next asks for a region.

> Yes, otherwise NextRegion will say "dunno what to give you: a collector is not running"

Why not just automatically start when the user asks for a region? Personally, I'd expect the API to be:

```go
qrb.SetLimit(N) // set the limit on the number of bytes to read relative to the current offset.
```

---

Basically, forget about the internals and try to figure out what you'd like the API to look like regardless of how this library is implemented under the hood.

Why would users care about explicitly starting/stopping the collector? Why does the user even care that there is a "collector".

The user cares about:

* Closing the buffer when done with it.
* Reading slices from the buffer.
* Setting limits (not actually sure _why_?).

File: qringbuf.go [hunk 14]
```diff
@@ -1,3 +1,191 @@
...
+//  • *Region.Bytes() is always a slice of the underlying buffer, no data copying takes place
+//  • Data backing a *Region is guaranteed to remain available / not be overwritten provided:
+//    ◦ NextRegion(…) has not been called again allowing further writes into the buffer
+//    ◦ *Region.Reserve() was invoked, which blocks writes until a *Region.Release()
+//
+// Examples
+//
+// In code the basic usage looks roughly like this:
+//
+//  qrb, initErr := qringbuf.NewFromReader( *someReader, qringbuf.Config{ … } )
```
[the remark this answers is not in the document]
Comment (ribasushi) [reply qringbuf.go]: I did and added the spaces afterwards, because of it rendering weird in a browser. The fact that you noticed this, means it is jarring for others too, so reverting all the post-formatting 😿 

https://github.com/ipfs/go-qringbuf/pull/6/commits/ee59439441fd

File: qringbuf.go (hunk 12 above)
Comment (ribasushi) [reply qringbuf.go]: I don't see how... that can work? We need to chat about this more...

An aside: I did have an implementation with sync.Pool in dagger itself over millions of 9-byte slices, where the freeing is definitive and thus a Pool makes sense -  I did rip it all out, because while allocations went down in a profile, there was **no actual difference, not even sub-%, outright undetectable**

Spoke with Eric a bit, it seems the compiler just got far enough along that sync.Pool is becoming needless overhead.

File: qringbuf.go (hunk 11 above)
Comment (ribasushi) [reply qringbuf.go]: We need to chat about this more...

At present the panic of "you gave me back everything I gave you" has saved me from countless endless-loops. Having to keep track of "what did I eat" and advancing your own "view" by hand seems... hrm.

As a general rule seeking back a super-short amount is common ( due to your last record having been cut ). Re-reading the entire buffer though is outright a bug.

File: qringbuf.go [hunk 15]
```diff
@@ -515,7 +787,14 @@ func (qrb *QuantizedRingBuffer) collector() {
...
 }
 
-func (qrb *QuantizedRingBuffer) Stop() (didResultInStop bool) {
+// StopFill is used to shutdown the collector goroutine without having reached
+// an io error on the underlying io.Reader. Calling this function blocks until
+// the collector goroutine exits. Note that if the goroutine is currently busy
+// performing a syscall as part of a Read(…), it may take an arbitrarily long
+// amount of time to return. In cases of a blocked system read you will need to
+// arrange for a signal to be delivered to your process, interrupting all
+// syscalls.
```
[the remark this answers is not in the document]
Comment (ribasushi) [reply qringbuf.go]: I guess I need to somehow make it clearer that a reader is not exclusive to the qringbuf, it merely cooperates. Will do so in the final commit to push.

I guess... what about changing doc above to say "if you want me to stop - retain and close the underlying reader", and removing the StopFill() method entirely ( thus reducing complexity way down ).

Actually **I love that**... here is how much crap this throws away: https://github.com/ipfs/go-qringbuf/commit/8eaa25fd2947

One extra doc update coming in a separate commit

[the remark this answers is not in the document]
Comment (ribasushi) [reply qringbuf.go]: And here is the doc part: https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L814-L817

File: qringbuf.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ribasushi) [reply qringbuf.go]: THANK YOU! This is exactly the type of review/comment-thread I was looking for. Sorry for rewriting half the docs - one would need to re-read this almost from scratch.

https://github.com/ipfs/go-qringbuf/commit/9f05222fd67 resolves all remaining comments ( I marked them as such ). Notably:

> Can we put this in the description for the individual methods? The reader won't know what any of these functions do till they read the actual API.

Re-wrapped, I did a bit of both. Worry is that the reader **won't read** down to the API without an overview making them go "huh... that's... not something I've seen before..."

> I'd focus more on why and what and less on how.

Tried to do so... not sure how successful. Let's chat about this more ;)

> Avoid exposing mutexes 

This is explicitly by design - expanded upon here: https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L8-L11 and here https://github.com/ipfs/go-qringbuf/blob/9f05222fd67da74107b91461c9f1d3c97e160ce3/qringbuf.go#L519-L521

File: qringbuf.go [hunk 16]
```diff
@@ -1,3 +1,211 @@
...
+by a standalone goroutine. It is primarily designed for processing a series
+of arbitrary sub-streams form a single io.Reader, each sub-stream in turn
+comprised of variable-length records.
+
+The buffer object DOES NOT ASSUME exclusive ownership of the supplied io.Reader,
+never reads more than instructed by an argument to StartFill(…), and exposes
+a standard sync.Mutex interface allowing pausing all operations when exclusive
+access of the underlying Reader is desired.
+
+Examples
```
Comment (abo-L) [inline qringbuf.go:13]: TON

**Review by abo-L — CHANGES_REQUESTED** [review]

**Review by abo-L — APPROVED** [review]
